db/crud.go

```diff
@@ -679,58 +655,95 @@ func (db *Database) Put(docid string, body Body) (newRevID string, doc *Document
+			if newDoc.RevID != "" {
 				// PUT with no parent rev given, but there is an existing current revision.
 				// This is OK as long as the current one is deleted.
-				if !doc.History[matchRev].Deleted {
+				if !doc.History[newDoc.RevID].Deleted {
 					return nil, nil, nil, base.HTTPErrorf(http.StatusConflict, "Document exists")
 				}
-				generation, _ = ParseRevID(matchRev)
+				generation, _ = ParseRevID(newDoc.RevID)
 				generation++
 			}
-		} else if !doc.History.isLeaf(matchRev) || db.IsIllegalConflict(doc, matchRev, deleted, false) {
+		} else if !doc.History.isLeaf(newDoc.RevID) || db.IsIllegalConflict(doc, newDoc.RevID, newDoc.Deleted, false) {
 			return nil, nil, nil, base.HTTPErrorf(http.StatusConflict, "Document revision conflict")
 		}
 
 		// Process the attachments, and populate _sync with metadata. This alters 'body' so it has to
 		// be done before calling createRevID (the ID is based on the digest of the body.)
-		newAttachments, err := db.storeAttachments(doc, newDoc.DocAttachments, generation, matchRev, nil)
+		newAttachments, err := db.storeAttachments(doc, newDoc.DocAttachments, generation, newDoc.RevID, nil)
 		if err != nil {
 			return nil, nil, nil, err
 		}
 
 		// Make up a new _rev, and add it to the history:
-		newRev, err := createRevID(generation, matchRev, newDoc.Body())
+		newRev, err := createRevID(generation, newDoc.RevID, newDoc.Body())
 		if err != nil {
 			return nil, nil, nil, err
 		}
 
-		if err := doc.History.addRevision(newDoc.ID, RevInfo{ID: newRev, Parent: matchRev, Deleted: deleted}); err != nil {
-			base.InfofCtx(db.Ctx, base.KeyCRUD, "Failed to add revision ID: %s, for doc: %s, error: %v", newRev, base.UD(docid), err)
+		if err := doc.History.addRevision(newDoc.ID, RevInfo{ID: newRev, Parent: newDoc.RevID, Deleted: newDoc.Deleted}); err != nil {
+			base.InfofCtx(db.Ctx, base.KeyCRUD, "Failed to add revision ID: %s, for doc: %s, error: %v", newRev, base.UD(newDoc.ID), err)
 			return nil, nil, nil, base.ErrRevTreeAddRevFailure
 		}
 
 		// move _attachment metadata to syncdata of doc after rev-id generation
 		doc.SyncData.Attachments = newDoc.DocAttachments
 		newDoc.RevID = newRev
-		newDoc.Deleted = deleted
+		newDoc.Deleted = newDoc.Deleted
 
 		return newDoc, newAttachments, nil, nil
 	})
 
+	return
+
+}
+
+// Updates or creates a document.
+// The new body's BodyRev property must match the current revision's, if any.
+func (db *Database) PutWithBody(docid string, body Body) (newRevID string, doc *Document, err error) {
+	// Get the revision ID to match, and the new generation number:
+	matchRev, _ := body[BodyRev].(string)
+	generation, _ := ParseRevID(matchRev)
+	if generation < 0 {
+		return "", nil, base.HTTPErrorf(http.StatusBadRequest, "Invalid revision ID")
+	}
+	deleted, _ := body[BodyDeleted].(bool)
+
+	expiry, err := body.ExtractExpiry()
+	if err != nil {
+		return "", nil, base.HTTPErrorf(http.StatusBadRequest, "Invalid expiry: %v", err)
+	}
```

Comment:
We should move the expiry and ParseRevID validation down into db.Put - so that both versions of Put are doing the checks. Here we should only be concerned about extracting the properties (not validating them)